### DISCUSSION TITLE
Can I use cpp programmer as a publisher and rust program as a subscriber?

### DISCUSSION THREAD

**Discussion by XiaoPengYouCode** [eclipse-iceoryx/iceoryx2#480] [Q&A] [answered]
I try to run the example of cpp publisher, and wanna to subscriber it on another rust project, with the same service name, but If I open the cpp service, I can open the rust service, the wrong message is under
```shell
Error: PublishSubscribeOpenError(IncompatibleTypes)
```

and I just run 
```
# first shell
./target/ffi/build/examples/cxx/publish_subscribe/example_cxx_publish_subscribe_publisher
#another shell
./target/debug/examples/publish_subscribe_subscriber
```

but all the type in the example is same
```rust
#[derive(Debug)]
#[repr(C)]
pub struct TransmissionData {
    pub x: i32,
    pub y: i32,
    pub funky: f64,
}
```
so how can I resolve this problem?
Thank you all so much!
-\(+_+)/-

**Comment by Chaoses-Ib** [comment]
Apart from the memory layout, iceoryx2 also requires the type name to be the same.

The Rust side uses `core::any::type_name()` to get the undecorated name, like `"std::option::Option<std::string::String>"`:

https://github.com/eclipse-iceoryx/iceoryx2/blob/a657228f03d9a7ed040616cedff940b43bd447a1/iceoryx2/src/service/static_config/message_type_details.rs#L44-L56

However, the C++ side uses `typeid().name`, whose format is undefined and left to the compiler. Acoording to [this question](https://stackoverflow.com/questions/4465872/why-does-typeid-name-return-weird-characters-using-gcc-and-how-to-make-it-prin), on GCC/Clang it's the mangled name, while on MSVC it's the demangled name. So the type name is not only different from Rust and C++, but also different from GCC/Clang and MSVC. (And sometimes even GCC and Clang will mangle the same name in different ways)

https://github.com/eclipse-iceoryx/iceoryx2/blob/a657228f03d9a7ed040616cedff940b43bd447a1/iceoryx2-ffi/cxx/include/iox2/message_type_details.hpp#L21-L42

Unfortunately, looks like there is currently no way in iceoryx2 to customize the type name.

**Comment by elfenpiff** [comment, marked as answer]
@XiaoPengYouCode @Chaoses-Ib 

With the introduction of the C++ and C language binding, we wanted to make it work first reliable in the language itself and then focus on cross language use-cases. So at the moment you can use iceoryx2 only for Rust 2 Rust or C/C++ to C/C++ communication. 

But by utilizing the internal Rust API you can establish a communication between C++ and Rust - it is internal and only a temporary solution!

```rust
// Rust part
struct TransmissionData {
  //...
}

let mut type_details = TypeDetail::__internal_new<TransmissionData>(TypeVariant::FixedSize);
type_details.type_name = "TransmissionData"; // set here the type name to the C++ compiler output of `typeid(TransmissionData).name()`

let service = node
    .service_builder(&"My/Funk/ServiceName".try_into()?)
    .publish_subscribe::<TransmissionData>()
    .__internal_set_payload_type_details(type_details)
    .open_or_create()?;
```

```cxx
// C++ part
struct TransmissionData {
  // ...
};
// this prints the value you have to use in the rust part type_details.type_name
std::cout << typeid(TransmissionData).name() << std::endl; 

// The C++ code can stay as it is since the Rust side has adjusted the type name to the C++ variant
```

The `typeid().name` and `core::any::type_name()` approach will soon be replaced by a `ShmSend` trait / derive macro approach on the Rust side and template inheritance on the C++ side.

So it will be looking somehow like this:

```rust
#[derive(ShmSend)]
#[type_details(TypeName = "MyFunkyType")]
struct MyFunkyType {
  a: u64,
  b: u64
}
```

```cxx
struct MyFunkyType {
  constexpr char * TYPE_NAME = "MyFunkyType";
  uint64_t a;
  uint64_t b;
};
```

Currently, we are preparing some conference talks for next week ROScon and OCX where we present iceoryx2 and right after that we will focus on this feature. So in the beginning of November you should be able to have Rust 2 C++/C communication with the official iceoryx2 API, without using the internal API.

**Reply by simon-acc** [reply]
Hey @elfenpiff , thank you for the thorough explanation :pray: ! I also played with the examples and tried to connect a C++ publisher to a Rust subscriber and ran into the issue of a type name mismatch (on `0.5.0`, released 23rd of December 2024). Just to rule out that I misunderstood something: This is not supposed to work yet and users who want to experiment with C++/Rust interoperability can use the temporary workaround you mentioned above, right?

**Reply by elfenpiff** [reply]
@simon-acc This is not yet implemented, but it is on my todo list - I hope that this gets finished until `0.6.0`

**Reply by Daanoz** [reply]
@elfenpiff How will this work when the payload is a byte array? Using the `__internal_set_payload_type_details` trick I can get it to work at the moment, but the type_name from the C++ publisher is `h`. So I guess this needs a similar trick to set the `TYPE_NAME`?
